src/test/scala-2/chiselTests/simulator/PeekPokeApiSpec.scala

```diff
@@ -0,0 +1,282 @@
+      val vDot = UInt((2 * w + vecDim - 1).W)
+    })
+  })
+
+  val a = io.in.bits.a
+  val b = io.in.bits.b
+
+  val result = Wire(chiselTypeOf(io.out.bits))
+
+  result.c :#= MuxCase(
+    0.U,
+    Seq(
+      (io.op === TestOp.Add) -> (a + b),
+      (io.op === TestOp.Sub) -> (a - b),
+      (io.op === TestOp.Mul) -> (a * b).take(w)
+    )
+  )
+
+  // Supress the following warning:
+  //   [W001] Casting non-literal UInt to chiselTests.simulator.TestPeekPokeEnum$CmpResult.
+  //   You can use chiselTests.simulator.TestPeekPokeEnum$CmpResult.safe to cast without this warning.
+  // The warning seems to be a (unrelated) bug
+  suppressEnumCastWarning {
+    result.cmp :#= Mux1H(
+      Seq(
+        (a < b) -> CmpResult.LT,
+        (a === b) -> CmpResult.EQ,
+        (a > b) -> CmpResult.GT
+      )
+    )
+  }
+
+  // addition of vectors
+  result.vSum :#= io.in.bits.v1.zip(io.in.bits.v2).map { case (x, y) => x +& y }
+  // inner product
+  result.vDot :#= io.in.bits.v1.zip(io.in.bits.v2).map { case (x, y) => x * y }.reduce(_ +& _)
+  // outer product
+  result.vOutProduct :#= io.in.bits.v1.map { x => VecInit(io.in.bits.v2.map { y => x * y }) }
+
+  io.out :#= Pipe(io.in.valid, result)
+}
+
+class PeekPokeAPISpec extends AnyFunSpec with ChiselSim with Matchers {
+  val rand = scala.util.Random
+
+  describe("PeekPokeAPI with testableData") {
+    val w = 32
+    it("should peek and poke various data types correctly") {
+      val numTests = 100
+      simulate(new TestPeekPokeEnum(w)) { dut =>
+        assert(w == dut.io.in.bits.a.getWidth)
+        val vecDim = dut.vecDim
+        val truncationMask = (BigInt(1) << w) - 1
+        for {
+          _ <- 0 until numTests
+          a = BigInt(w, rand)
+          b = BigInt(w, rand)
+          v1 = Seq.fill(vecDim)(BigInt(w, rand))
+          v2 = Seq.fill(vecDim)(BigInt(w, rand))
+          op <- TestOp.all
+        } {
+
+          dut.io.in.bits.poke(
+            chiselTypeOf(dut.io.in.bits).Lit(
+              _.a -> a.U,
+              _.b -> b.U,
+              _.v1 -> Vec.Lit(v1.map(_.U(w.W)): _*),
+              _.v2 -> Vec.Lit(v2.map(_.U(w.W)): _*)
+            )
+          )
+          dut.io.in.valid.poke(true)
+          dut.io.op.poke(op)
+          dut.clock.step()
+          dut.io.in.valid.poke(false)
+
+          val peekedOp = dut.io.op.peek()
+          assert(peekedOp.litValue == op.litValue)
+          assert(peekedOp.toString.contains(TestOp.getClass.getSimpleName.stripSuffix("$")))
+
+          val expected = op match {
+            case TestOp.Add => a + b
+            case TestOp.Sub => a - b
+            case TestOp.Mul => a * b
+            case _          => throw new Exception("Invalid operation")
+          }
+          val expectedCmp = a.compare(b) match {
+            case -1 => dut.CmpResult.LT
+            case 0  => dut.CmpResult.EQ
+            case 1  => dut.CmpResult.GT
+          }
+
+          dut.io.out.valid.expect(true.B)
```

Comment:
Would it be possible to split this into a bunch of smaller tests of just individual types? Having the peeks and pokes of all types together kind of muddles what's being tested a bit.

Reply:
I think the most reliable way to test peeks and pokes is to apply `poke()`s to the inputs, let the circuit do its thing, and then `peek()` at or `expect()` the outputs. I really think we need even more testing/coverage, not less.

Reply:
I'm not saying have less testing or coverage, I'm just saying to do it in smaller chunks. For example, one could organize this by type and have basically a type parametered passthrough module (or have a register in there to delay a cycle), and then for each type and peek and poke method have a test that pokes the input and then checks the output.

My only real qualm is the "various" in `it("should peek and poke various data types correctly")`, I'd rather see:
```scala
it("should peek and poke Bool correctly")
it("should peek and poke UInt correctly")
it("should peek and poke ChiselEnum correctly")
... etc.
```
I'm not going to sandbag over this, I just think that would be better

Reply:
Oh I see what you mean and do agree with you. I cleaned up the tests as much as I could. Please let me know if you think it still needs more cleanup.